src/run.rs

```diff
@@ -0,0 +1,180 @@
+
+        ensure!(
+            command.status().context("Could not start git")?.success(),
+            "Cannot create working tree"
+        );
+    } else {
+        let mut command = process::Command::new("cargo");
+        command
+            .current_dir(&tmp_dir)
+            .args(["init", "--name", project_name.as_str()]);
+
+        if lib {
+            command.arg("--lib");
+        }
+
+        ensure!(
+            command.status().context("Could not start cargo")?.success(),
+            "Cargo command failed"
+        );
+    }
+
+    Ok(tmp_dir)
+}
+
+pub fn add_dependencies_to_project(tmp_dir: &Path, dependencies: &[Dependency]) -> Result<()> {
+    let mut toml = fs::OpenOptions::new()
+        .append(true)
+        .open(tmp_dir.join("Cargo.toml"))?;
+    for dependency in dependencies.iter() {
+        match dependency {
+            Dependency::CrateIo {
+                name: s,
+                version: v,
+            } => match &v {
+                Some(version) => writeln!(toml, "{} = \"{}\"", s, version)?,
+                None => writeln!(toml, "{} = \"*\"", s)?,
+            },
+            Dependency::Repository {
+                name,
+                url,
+                branch,
+                rev,
+            } => {
+                write!(toml, "{name} = {{ git = {url:?}", name = name, url = url)?;
+                if let Some(branch) = branch {
+                    write!(toml, ", branch = {:?}", branch)?;
+                }
+                if let Some(rev) = rev {
+                    write!(toml, ", rev = {:?}", rev)?;
+                }
+                writeln!(toml, " }}")?;
+            }
+        }
+    }
+
+    Ok(())
+}
+
+pub fn generate_delete_file(tmp_dir: &Path) -> Result<PathBuf> {
+    let delete_file = tmp_dir.join("TO_DELETE");
+    fs::write(
+        &delete_file,
+        "Delete this file if you want to preserve this project",
+    )?;
+
+    Ok(delete_file)
+}
+
+pub fn start_shell(config: &Config, tmp_dir: &Path) -> Result<()> {
+    let mut shell_process = match config.editor {
+        None => process::Command::new(get_shell()),
+        Some(ref editor) => {
+            let mut ide_process = process::Command::new(editor);
+            ide_process
+                .args(config.editor_args.iter().flatten())
+                .arg(tmp_dir);
+            ide_process
+        }
+    };
+
+    if env::var("CARGO_TARGET_DIR").is_err() {
+        if let Some(path) = &config.cargo_target_dir {
+            shell_process.env("CARGO_TARGET_DIR", path);
+        }
+    }
+
+    shell_process
+        .current_dir(&tmp_dir)
+        .status()
+        .context("Cannot start shell")?;
+
+    #[cfg(windows)]
+    if config.editor.is_some() {
+        unsafe {
+            cargo_temp_bindings::Windows::Win32::SystemServices::FreeConsole();
+        }
+    }
+
+    Ok(())
+}
+
+pub fn clean_up(
+    delete_file: PathBuf,
+    tmp_dir: TempDir,
+    worktree_branch: Option<Option<String>>,
```

Comment:
You don't need ownership.
```suggestion
    worktree_branch: Option<&Option<String>>,
```